File: Python/ByRound/1841/1841_E_Fill_the_Matrix.py
# Problem: CF 1841 E - Fill the Matrix
# https://codeforces.com/contest/1841/problem/E

"""
Algorithm: Dynamic Programming + Stack + Greedy
Time Complexity: O(n)
Space Complexity: O(n)

The problem involves placing integers in a matrix such that:
- Black cells are not filled
- Each row has at most one integer
- Beauty = count of consecutive integers in same row

Approach:
1. For each column i, a[i] black cells (from top) and (n - a[i]) white cells
2. Use a stack-based method to compute how many valid placements are possible
   by processing columns and calculating the 'gap' between black cell heights
3. Precompute possible beauty contributions for each gap size
4. Greedily assign integers to maximize beauty

The code computes the maximum possible beauty by:
1. Processing columns from left to right using a stack
2. For each such column, calculating how many integers can be placed in
   a given row to get a "jump" of a certain size (beauty value)
3. Using reverse processing also to handle transitions from right to left
4. Finally assigning the integers greedily based on precomputed values
"""

import sys
from array import array

input = lambda: sys.stdin.buffer.readline().decode().rstrip()
inp = lambda dtype: [dtype(x) for x in input().split()]
debug = lambda *x: print(*x, file=sys.stderr)
sum_n = lambda n: (n * (n + 1)) // 2
get_bit = lambda x, i: (x >> i) & 1
ceil_ = lambda a, b: a // b if (a >= 0) ^ (b > 0) else (abs(a) + abs(b) - 1) // abs(b)
Mint, Mlong, out = 2**30 - 1, 2**62 - 1, []


def solve(a, eq):
    stk = array("i") # Stack to keep track of column indices

    for i in range(n):
        ma = 0
        # While stack is not empty and current height is greater than stack top
        while stk and a[i] > a[stk[-1]]:
            ma = max(ma, a[stk.pop()]) # Update max height seen in popped elements

        # If stack is not empty and we're not dealing with equal heights (if eq == 0)
        # Or if there's an actual inequality in heights
        if stk and (a[i] != a[stk[-1]] or eq == 0):
            val = max(i - stk[-1] - 2, 0) # Calculate gap
            # Add contribution to number of possible placements
            mem[val] += (a[i] - ma) * val
        elif not stk and i: # If stack is empty and not the first column
            mem[i - 1] += (a[i] - ma) * (i - 1)

        stk.append(i) # Push the column index onto the stack


for _ in range(1, int(input()) + 1):
    n = int(input())
    a = array("i", inp(int))
    m, ans = int(input()), 0
    mem = [0] * n # Array to store how many placements give a certain beauty
    mem[n - 1] = (n - max(a)) * (n - 1) # Base case initialization
    
    solve(a, 0) # Forward pass
    solve(a[::-1], 1) # Reverse pass to handle other directions

    # Greedily assign integers to maximize beauty
    for i in range(n - 1, 0, -1):
        mi = min(m, mem[i] + ceil_(mem[i], i))
        m -= mi
        div, mod = divmod(mi, i + 1)
        ans += div * i + max(0, mod - 1)

    out.append(ans)

print("\n".join(map(str, out)))


# https://github.com/VaHiX/CodeForces/

File: CPP/ByTier/E/1841_E_Fill_the_Matrix.cpp
// Problem: CF 1841 E - Fill the Matrix
// https://codeforces.com/contest/1841/problem/E

/*
Algorithm: Monotonic Stack + Greedy
Approach:
- The problem involves placing integers in a matrix such that:
  - Black cells cannot contain integers.
  - Beauty is defined as the number of adjacent pairs (j, j+1) in the same row.
- For each column, we know how many black cells there are (a[i]), so the rest are white.
- Using a monotonic stack, we precompute the number of white cells in each "segment"
  that can be filled greedily to maximize beauty.
- Then, we use a greedy strategy to assign the integers, maximizing beauty at each step.

Time Complexity: O(n) amortized per test case due to stack operations and linear traversal.
Space Complexity: O(n) for storing arrays and the stack.

*/
#include <algorithm>
#include <iostream>
#include <stack>
#include <vector>

#define endl '\n'
using namespace std;
using ll = long long;
int MOD = 114514;
signed main() {
  ios::sync_with_stdio(0), cin.tie(0), cout.tie(0);
  int T;
  cin >> T;
  while (T--) {
    int n;
    cin >> n;
    vector<int> a(n + 2);
    a[0] = a[n + 1] = n;  // Sentinel values for easier computation
    for (int i = 1; i <= n; ++i)
      cin >> a[i];
    vector<ll> suf(n + 1);  // suf[i] stores the number of white cells in segments of width i
    stack<int> st;
    st.push(0);
    for (int i = 1; i <= n + 1; ++i) {
      // Maintain a decreasing stack based on a[i]
      while (st.size() >= 2 && a[st.top()] < a[i]) {
        int k = st.top();
        st.pop();
        int j = st.top();
        // Compute contribution of the segment (j+1, k) to suf[i - j - 1]
        suf[i - j - 1] += min(a[i], a[j]) - a[k];
      }
      st.push(i);
    }
    ll m;
    cin >> m;
    ll ans = 0;
    // Greedily assign integers to maximize beauty
    for (int i = n; i; --i) {
      // Determine how many full segments of size i we can fill
      ll mi = min(suf[i], m / i);
      ans += mi * (i - 1);  // Add beauty from these segments
      suf[i - 1] += suf[i] - mi;  // Propagate remaining segments
      m -= mi * i;  // Reduce remaining integers
    }
    cout << ans << endl;
  }
  return 0;
}


// https://github.com/VaHiX/CodeForces/